### DISCUSSION TITLE
Optimised API for  guarding double addition and atomic modification to delegates / events

### DISCUSSION THREAD

**Discussion by sirinath** [dotnet/runtime#63118] [Ideas]
Can API be added so programmers can detect if an function is already added and/or add only if it is not already present.

One can do something like this. But perhaps more performance can be got if there is a way to see if it is already added and / or add if not already added.

```csharp
          ticks.Tick -= NewTicksInit;
          ticks.Tick += NewTicksInit;
```

Also and atomic way  change event / delegate handle (remove old handle if present and add new handle if not already added) would be helpful without using locks. E.g. atomic form of which does not loose events:

```csharp
          ticks.Tick -= OldTickArrivedInit;
          ticks.Tick -= NewTickArrivedInit; // incase already added
          ticks.Tick += NewTickArrivedInit;
```

This can be extended to multiple atomic removals followed by additions with and without checks to see if the handles are already present.

**Comment by huoyaoyuan** [comment]
Currently a single +=/-= operation is implemented in atomic way, with existence check inside it. Can you express your usage with lock statement, so that we can discuss about improving it?

**Comment by sirinath** [comment]
> Can you express your usage with lock statement, so that we can discuss about improving it?

Broadly to add remove multiple listeners in an atomic transaction.

>  with existence check inside it.

I am not sure if an existence check is done. There many SO answers which suggest otherwise. E.g. https://stackoverflow.com/questions/136975/has-an-event-handler-already-been-added

**Reply by huoyaoyuan** [reply]
You can just write the usage and inspect it in sharplab
